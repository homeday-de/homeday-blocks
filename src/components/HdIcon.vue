<template>
  <inline-svg
    v-bind="$attrs"
    :transform-source="transform"
  />
</template>

<script>
import _intersection from 'lodash/intersection';
import InlineSvg from 'vue-inline-svg';

export default {
  name: 'HdIcon',
  inheritAttrs: false,
  components: {
    InlineSvg,
  },
  props: {
    rotate: {
      type: Number,
      default: 0,
    },
    fillFromClass: {
      type: Object,
      default: null,
    },
    // To add class
    classFromFill: {
      type: Object,
      default: null,
    },
  },
  methods: {
    transform(svg) {
      if (!this.fillFromClass && !this.classFromFill) {
        return svg;
      }

      const paths = svg.querySelectorAll('path');
      paths.forEach((path) => {
        const newFill = this.getFillFromClassNames(path.getAttribute('class'));
        if (newFill) {
          path.setAttribute('fill', newFill);
        }

        const newClassName = this.getClassNameFromFill(path.getAttribute('fill'));
        if (newClassName) {
          path.setAttribute('class', `${path.getAttribute('class')} ${newClassName}`);
        }
      });

      return svg;
    },
    getFillFromClassNames(classNames) {
      if (!classNames || !this.fillFromClass) {
        return null;
      }

      const classNameMatches = _intersection(
        (classNames || '').split(' '),
        Object.keys(this.fillFromClass),
      );

      return this.fillFromClass[classNameMatches[0]];
    },
    getClassNameFromFill(fill) {
      if (!fill || !this.classFromFill) {
        return '';
      }

      return this.classFromFill[fill];
    },
  },
};
</script>

<style lang="scss" scoped>

</style>
