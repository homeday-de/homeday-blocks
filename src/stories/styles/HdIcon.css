svg:hover .secondary-color {
  fill: coral;
}

b {
  font-weight: bold;
}