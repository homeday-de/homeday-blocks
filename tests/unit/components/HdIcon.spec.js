import _merge from 'lodash/merge';
import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import InlineSvg from 'vue-inline-svg';
import HdIcon from '@/components/HdIcon.vue';

const ICON_CONTENT = `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" viewBox="222 126 53 53" width="50" height="50">
  <path class="color1" d="M272.48 137.62L258.61 150.5L244.75 141.58L254.65 126.73L272.48 137.62Z" fill="#1895FF"/>
  <path class="color2" d="M242.22 176.24L222.97 160.68L230.47 145.1L243.56 155.68L242.22 176.24Z" fill="#1C3553"/>
</svg>`;

const stubbedInlineSvg = _merge(InlineSvg, {
  methods: {
    download() {
      const parser = new DOMParser();
      const result = parser.parseFromString(ICON_CONTENT, 'text/xml');
      const svgEl = result.querySelector('svg');
      return Promise.resolve(this.transformSource(svgEl));
    },
  },
});

describe('HdIcon', () => {
  const wrapperFactory = wrapperFactoryBuilder(HdIcon, {
    propsData: {
      src: 'fake/icon1.svg',
    },
    stubs: {
      InlineSvg: stubbedInlineSvg,
    },
  });

  let wrapper;
  beforeEach(() => {
    wrapper = wrapperFactory();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('renders as expected', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('has an <svg> in the root', () => {
    expect(wrapper.is('svg')).toBe(true);
  });

  it('sets the right fill based on `fillFromClass`', async () => {
    const TEST_CLASS = 'color1';
    const TEST_FILL = 'orange';

    wrapper = wrapperFactory({
      propsData: {
        src: 'fake/icon2.svg',
        fillFromClass: {
          [TEST_CLASS]: TEST_FILL,
        },
      },
    });

    await wrapper.vm.$nextTick();

    const targetedPaths = wrapper.findAll(`.${TEST_CLASS}`);

    expect(targetedPaths.length).toBeGreaterThan(0);
    expect(targetedPaths.is(`[fill="${TEST_FILL}"]`)).toBe(true);
  });

  it('sets the right class based on `classFromFill`', async () => {
    const TEST_FILL = '#1C3553';
    const TEST_CLASS = 'primary-color';

    wrapper = wrapperFactory({
      propsData: {
        src: 'fake/icon3.svg',
        classFromFill: {
          [TEST_FILL]: TEST_CLASS,
        },
      },
    });

    await wrapper.vm.$nextTick();

    const targetedPaths = wrapper.findAll(`[fill="${TEST_FILL}"]`);

    expect(targetedPaths.length).toBeGreaterThan(0);
    expect(targetedPaths.is(`.${TEST_CLASS}`)).toBe(true);
  });
});
