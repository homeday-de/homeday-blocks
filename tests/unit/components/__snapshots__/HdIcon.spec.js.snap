// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdIcon renders as expected 1`] = `
<svg height="50" width="50" viewBox="222 126 53 53" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns="http://www.w3.org/2000/svg">
  <path class="color1" d="M272.48 137.62L258.61 150.5L244.75 141.58L254.65 126.73L272.48 137.62Z" fill="#1895FF"></path>
  <path class="color2" d="M242.22 176.24L222.97 160.68L230.47 145.1L243.56 155.68L242.22 176.24Z" fill="#1C3553"></path>
</svg>
`;
